### PR TITLE
FEATURE - Home Screen Quick Actions

### LINKED ISSUE #65: FEATURE - Home Screen Quick Action
## Describe the feature you want the app to have that it currently doesn't
* Create a [Home Screen Quick Action](https://developer.apple.com/documentation/uikit/menus_and_shortcuts/add_home_screen_quick_actions) that will open the `AddMaintenanceView` to easily add a maintenance event

## Describe alternatives you've considered
* N/A

## Additional context
* Possibly work on how you can make this code testable and easy to use later on App Intents and Siri Shortcuts

### PR DESCRIPTION
**Pull request by OmarHegazy93**
# What it Does
Implementing Home Screen quick actions for Add maintenance in [issue-65](https://github.com/mikaelacaron/Basic-Car-Maintenance/issues/65) as well as Add new vehicle
* Closes #65
## Implementing Home quick actions using the following flow:
1. Receive the shortcutItem when the is being launched or when the app is already running in background
2. Publish the new value using the singleton `ActionService` class by injecting it as an EnvironmentObject to all views
3. The `MainTabView` will be responsible for selecting the corresponding tab to make sure the view is presented from the relevant hierarchy
4. Any child view will that has a service will be responsible for dismissing its presented view (if it's not the desired one) and to show it if it is (To prevent presenting a view while another view is already presented)

# How I Tested
### Scenario 1:
1. Open the App
2. move it to background
3. long press on the app icon
4. select "**Add Maintenance**"
5. The "**Add Maintenance**"  screen should be opened once the app is moved to foreground

### Scenario 2:
1. Open the App
2. Move it to background
3. Long press on the app icon
4. Select "**Add Vehicle**"
5. The "**Add Vehicle**"  screen should be opened  once the app is moved to foreground

### Scenario 3:
1. Open the App
2. Move it to background
3. Long press on the app icon
4. Select "**Add Maintenance**"
5. When the opens and "**Add Maintenance**" screen appears, move the app to background again
6. Long press on the app icon
7. Select "**Add Vehicle**"
8. The "**Add Maintenance**" screen  should be dismissed and the "**Add Vehicle**"  screen should be opened

### Scenario 4:
1. Open the App
2. Move it to background
3. Long press on the app icon
4. Select "**Add Vehicle**"
5. When the opens and "**Add Vehicle**" screen appears, move the app to background again
6. Long press on the app icon
7. Select "**Add Maintenance**"
8. The "**Add Vehicle**" screen should be dismissed and the "**Add Maintenance**" screen should be opened

### Scenario 5:
1. When the app is terminated, long press on the app icon
2. Select "**Add Maintenance**" or "**Add Vehicle**"
3. The selected service should appear once the app is opened

# Notes
* Anything else that should be noted about how you implemented this feature?

# Screenshot
* Add a screenshot of your new feature! OR show a screen recording of it in action. (On the simulator press Cmd + R to record, and the stop button when done). **This video should be no longer than 30 seconds.**

### CODE REVIEW

File: Basic-Car-Maintenance/Info.plist [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>UIApplicationShortcutItems</key>
+	<array>
+		<dict>
+			<key>UIApplicationShortcutItemIconSymbolName</key>
+			<string>wrench.and.screwdriver.fill</string>
+			<key>UIApplicationShortcutItemSubtitle</key>
+			<string>Add new maintenence service</string>
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Info.plist]: Can you remove the subtitle in both actions?

File: Basic-Car-Maintenance/Shared/BasicCarMaintenanceApp.swift [hunk 2]
```diff
@@ -10,27 +10,66 @@ import SwiftUI
...
-                SettingsView(authenticationViewModel: authenticationViewModel)
-                    .tabItem {
-                        Label("Settings", systemImage: "gear")
-                    }
+      private let actionService = ActionService.shared
+      @StateObject var authenticationViewModel: AuthenticationViewModel
+      @UIApplicationDelegateAdaptor(AppDelegate.self) var delegate
+      
+      init() {
+            FirebaseApp.configure()
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/BasicCarMaintenanceApp.swift]: can you move this into this function in the AppDelegate

```swift
func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey : Any]? = nil) -> Bool {
}
```

File: Basic-Car-Maintenance/Shared/Models/ActionService.swift [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+//
+
+import UIKit
+
+enum ActionType: String {
+      case newMaintenance = "NewMaintenance"
+      case addVehicle = "AddVehicle"
+}
+
+enum Action: Equatable {
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Models/ActionService.swift]: Can you fix the indentation on this?
Use `4` spaces, and the shortcut `Ctrl + i`

File: Basic-Car-Maintenance/Shared/Models/ActionService.swift [hunk 4]
```diff
@@ -0,0 +1,31 @@
+//
+//  ActionService.swift
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Models/ActionService.swift]: Can you rename this file to `Action` rather than `ActionService`

File: Basic-Car-Maintenance/Shared/Settings/Views/SettingsView.swift [hunk 5]
```diff
@@ -136,6 +137,26 @@ struct SettingsView: View {
...
+                // so we will need to dismiss the current presented view
+                isShowingAddVehicle = false
+                return
+            }
+            
+            // if the view is already presented, do nothing
+            guard !isShowingAddVehicle else { return }
+            // delay the presentation of the view a bit
+            // to make sure the already presented view is dismissed.
+            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Settings/Views/SettingsView.swift]: Can you change this to a delay of `0.1` to `0.3` rather than `0.5`?

File: Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift [hunk 6]
```diff
@@ -92,6 +84,41 @@ struct DashboardView: View {
...
+                // another action has been triggered, so we will need to dismiss the current presented view
+                isShowingAddView = false
+                return
+            }
+            
+            // if the view is already presented, do nothing
+            guard !isShowingAddView else { return }
+            // delay the presentation of the view a bit
+            // to make sure the already presented view is dismissed.
+            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift]: Same comment about making this delay shorter

File: Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift [hunk 7]
```diff
@@ -92,6 +84,41 @@ struct DashboardView: View {
...
+                isShowingAddView = true
+            }
+        }
+    }
+    
+    private func showAddMaintenanceView() -> some View {
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift]: Change this function name to `makeAddMaintenanceView`

File: Basic-Car-Maintenance/Shared/Main/MainTabView.swift [hunk 8]
```diff
@@ -0,0 +1,57 @@
+//
+//  MainTabView.swift
+//  Basic-Car-Maintenance
+//
+//  Created by Omar Hegazy on 05/10/2023.
+//
+
+import SwiftUI
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Main/MainTabView.swift]: Can you fix the indentation, using `4` spaces, please

File: Basic-Car-Maintenance/Shared/Main/MainTabView.swift [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+      case settings = 1
+}
+
+struct MainTabView: View {
+      @EnvironmentObject var actionService: ActionService
+      @Environment(\.scenePhase) var scenePhase
+      @StateObject var authenticationViewModel = AuthenticationViewModel()
+      @State var selectedTab: TabSelection = .dashboard
+      @State var isShowingAddVehicle = false
+      @State var isShowingAddView = false
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Main/MainTabView.swift]: These can be deleted, they're not used anywhere

File: Basic-Car-Maintenance.xcodeproj/project.pbxproj [hunk 10]
```diff
@@ -268,6 +282,7 @@
 				FFC8CDA32AA385E800D129A6 /* GoogleService-Info.plist */,
 				FF5D13A62A86C2D600BC9BD6 /* BasicCarMaintenanceApp.swift */,
 				FFBFE08F2A98EFDD000A9BEB /* Models */,
+				8AEE81702ACF383200FC0C2A /* Main */,
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance.xcodeproj/project.pbxproj]: Can you remove the `Main` group, and instead put the `MainTabView` file right under the `BasicCarMaintenanceApp` file?

**Review by mikaelacaron — CHANGES_REQUESTED** [review]
There's one issue: click Add, it doesn't dismiss the sheet, for the maintenance event, but it did for the add vehicle

Just a few comments, and we talked on my [livestream](https://youtube.com/live/h2Xzk41_8Xs?feature=share)

Great work!!

**Comment by mikaelacaron** [comment]
@OmarHegazy93 pull changes again, I just fixed all the signing issue stuff. I'm sorry there are most likely going to be merge conflicts! Thanks for all your work on this!

File: Basic-Car-Maintenance.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved [hunk 11]
```diff
@@ -1,113 +0,0 @@
-{
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved]: Why is this file deleted?

File: Basic-Car-Maintenance/Info.plist [hunk 12]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Info.plist]: Should this be at the top level? Or can it be defined inside the `Shared` folder?

If it can be in the `Shared` folder, can you put it right under the `GoogleService-Info.plist`
Screenshot 2023-10-07 at 4 21 02 PM

**Review by mikaelacaron — CHANGES_REQUESTED** [review]
Great work! I have a question about the `Info.plist`

Please merge from `dev` into your branch? There are going to be merge conflicts, unfortunately

Once you've made the changes, please click the re-review button and I'll be notified, thanks!
Re-review button

File: Basic-Car-Maintenance.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved (hunk 11 above)
Comment (OmarHegazy93) [reply Basic-Car-Maintenance.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved]: looks like was done by mistake from last merge, should be good now 👍

**Review by mikaelacaron — CHANGES_REQUESTED** [review]
@OmarHegazy93 There's still a conflict in the `pbxproj` file

**Review by mikaelacaron — CHANGES_REQUESTED** [review]
Thanks for your work on this! Last thing, is project now uses the new observation framework with the `@Observable` macro, can you please convert your code to using that

This article is helpful!

File: Basic-Car-Maintenance/Shared/BasicCarMaintenanceApp.swift [hunk 13]
```diff
@@ -10,26 +10,68 @@ import SwiftUI
 
 @main
 struct BasicCarMaintenanceApp: App {
-    @State private var authenticationViewModel: AuthenticationViewModel
+    private let actionService = ActionService.shared
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/BasicCarMaintenanceApp.swift]: Is there a specific reason this is a singleton?

**Review by mikaelacaron — APPROVED** [review]
Great work!! Thanks for continually updating the PR through merge conflicts
I'm committing some changes to use the new observation framework
